README
Dist::Zilla::App::Command::critic

"critic" is an "App::Command" for "Dist::Zilla" which streamlines
running "Perl::Critic" on your built distribution.

This competes with the likes of "[Test::Perl::Critic]" by:

*   not requiring the rest of the steps in the test life-cycle to
    execute.

*   not being impeded by the other tests cluttering your output.

*   not suffering the limitations of "Test::Perl::Critic" which discards
    profile color settings.

*   carefully formatting output to give a clearer visualization of where
    failures lie.

*   not requiring your dist have a "Test::Perl::Critic" test pass for
    release.

*   not requiring your dist to have any explicit "Perl::Critic"
    consumption.

Behaviorally:

  dzil critic

Behaves very similar to:

   dzil run --no-build perlcritic -p perlcritic.rc lib/

Except with improved verbosity of file name reporting.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Dist-Zilla-App-Command-critic

0.001012 2017-03-04T04:36:10Z d511b37

0.001011 2017-03-04T04:13:16Z dbde69e
 [Bugfix]
 - Avoid test failures on 5.26 w/ -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/Dist-Zilla-App-Command-critic/issues/1

 [Dependencies::Stats]
 - Dependencies changed since 0.001010, see misc/*.deps* for details
 - develop: +5 ↑2 -1 (suggests: ↑2)
 - test: +2 ↓1 -1

0.001010 2015-04-12T20:14:30Z 237f1a1
 [Dependencies::Stats]
 - Dependencies changed since 0.001004, see misc/*.deps* for details
 - develop: +3 ↑3 -1 (recommends: ↓1, suggests: ↑2)
 - runtime: (recommends: +1)
 - test: ↑1 -1 (recommends: ↓1)

 [Minor]
 - Moving app testing to dztest.
 - DESCRIPTION reworked.

0.001004 2014-11-01T13:07:20Z 437dbe4
 [Bugfix]
 - Avoid test failures due to old versions of Test::More not doing subtest right ( via upgrade on the module that
   uses it )

 [Dependencies::Stats]
 - Dependencies changed since 0.001003, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - test: ↑1

0.001003 2014-10-30T23:33:56Z ce4d03b
 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +3 -1 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1
 - test: +1 (recommends: ↑2)

 [Internals]
 - Stop loading utf8 ( optimisation )
 - 5.006 is now minimum perl.

0.001002 2014-10-12T14:14:11Z 8fd643f
 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - configure: ↓1 (recommends: ↑1)
 - develop: +8 ↑2 -3 (suggests: +2 -1)
 - test: +5 -3 (recommends: ↑3 -1)

 [Meta]
 - Move to KENTNL github org.
 - Author = kentnl@cpan

 [Packaging]
 - Packaging cleaned up a bit

 [Tests]
 - Basic behaviour tests added

0.001001 2014-04-13T01:46:56Z
 [Bugfix]
 - Include bin and script in all_perl_files search roots. I'm not sure how Test::Perl::Critic seems to do this.

0.001000 2014-04-12T23:42:06Z
 - First version.

 [Dependencies::Aggregate]
 - 5.008: 216
 - 5.010: 196
 - 5.019: 175


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

# $bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
# $bundle->remove_policy('CodeLayout::RequireUseUTF8');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "build your dist and run Perl::Critic on the built files.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-App-Command-critic",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::App::Command::critic",
  "PREREQ_PM" => {
    "Dist::Zilla::App" => 0,
    "Path::Tiny" => 0,
    "Perl::Critic" => 0,
    "Perl::Critic::Utils" => 0,
    "Perl::Critic::Utils::Constants" => 0,
    "Perl::Critic::Violation" => 0,
    "Try::Tiny" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dist::Zilla::App::Tester" => 0,
    "Dist::Zilla::Plugin::GatherDir" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::DZil" => 0,
    "Test::More" => 0,
    "Test::TempDir::Tiny" => 0
  },
  "VERSION" => "0.001012",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Dist::Zilla::App" => 0,
  "Dist::Zilla::App::Tester" => 0,
  "Dist::Zilla::Plugin::GatherDir" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Path::Tiny" => 0,
  "Perl::Critic" => 0,
  "Perl::Critic::Utils" => 0,
  "Perl::Critic::Utils::Constants" => 0,
  "Perl::Critic::Violation" => 0,
  "Test::DZil" => 0,
  "Test::More" => 0,
  "Test::TempDir::Tiny" => 0,
  "Try::Tiny" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
